tasks/libs/ciproviders/gitlab_api.py

```diff
@@ -84,25 +114,108 @@ def __init__(self, references):
+        return YamlReferenceTagList(loader.construct_sequence(node))
 
     @classmethod
     def to_yaml(cls, dumper, data):
-        return dumper.represent_sequence(cls.yaml_tag, data.data, flow_style=True)
+        return dumper.represent_sequence(cls.yaml_tag, data, flow_style=True)
+
+
+def convert_to_config_node(node):
+    """
+    Convert json like data to ConfigNode
+    """
+    if isinstance(node, dict):
+        return ConfigNodeDict({k: convert_to_config_node(v) for k, v in node.items()})
+    elif isinstance(node, list):
+        constructor = YamlReferenceTagList if isinstance(node, YamlReferenceTagList) else ConfigNodeList
+
+        return constructor([convert_to_config_node(v) for v in node])
+    else:
+        return node
 
 
-def generate_gitlab_full_configuration(input_file, context=None, compare_to=None):
+def apply_yaml_extends(config: dict, node):
+    """
+    Applies `extends` yaml tags to the node and its children inplace
+    """
+    if not isinstance(node, dict):
+        return
+
+    if 'extends' in node:
+        parents = node['extends']
+        if isinstance(parents, str):
+            parents = [parents]
+
+        # Merge parent
+        for parent_name in parents:
+            parent = config[parent_name]
+            apply_yaml_postprocessing(config, parent)
+            for key, value in parent.items():
+                if key not in node:
+                    node[key] = value
+
+        del node['extends']
+
+
+def apply_yaml_reference(config: dict, node):
+    """
+    Applies `!reference` gitlab yaml tags to the node and its children inplace
+    """
+
+    def apply_ref(value):
+        """
+        Applies reference tags
+        """
+        if isinstance(value, YamlReferenceTagList):
+            assert value != [], 'Empty reference tag'
+
+            # [a, b, c] -> config[a][b][c]
+            ref_value = config[value[0]]
+            for i in range(1, len(value)):
+                ref_value = ref_value[value[i]]
+
+            apply_yaml_postprocessing(config, ref_value)
+
+            return ref_value
+        else:
+            apply_yaml_postprocessing(config, value)
+
+            return value
+
+    if isinstance(node, dict):
+        for key, value in node.items():
+            node[key] = apply_ref(value)
+    elif isinstance(node, list):
+        for i, value in enumerate(node):
+            node[i] = apply_ref(value)
+
+
+@lru_cache(maxsize=None)
+def apply_yaml_postprocessing(config: ConfigNodeDict, node):
+    if isinstance(node, dict):
+        for value in node.values():
+            apply_yaml_postprocessing(config, value)
+    elif isinstance(node, list):
+        for value in node:
+            apply_yaml_postprocessing(config, value)
+
+    apply_yaml_extends(config, node)
+    apply_yaml_reference(config, node)
+
+
+def generate_gitlab_full_configuration(
+    input_file, context=None, compare_to=None, return_dump=True, apply_postprocessing=False
+):
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏Please add the different args use description in the function docstring